eip712/src/main/java/pm/gnosis/eip712/EIP712.kt

```diff
@@ -0,0 +1,51 @@
+package pm.gnosis.eip712
+
+import pm.gnosis.crypto.utils.Sha3Utils
+import pm.gnosis.model.Solidity
+import pm.gnosis.model.SolidityBase
+import pm.gnosis.utils.hexToByteArray
+
+val solidityTypes = Solidity.types.keys
+const val DOMAIN_TYPE = "EIP712Domain"
+
+sealed class Type712 {
+    abstract val type: String
+    abstract fun encodeData(): ByteArray
+}
+
+class Literal712(val value: SolidityBase.Type) : Type712() {
+    override fun encodeData(): ByteArray = when (value) {
+        // Solidity.String is also Solidity.Bytes
+        is Solidity.Bytes -> Sha3Utils.keccak(value.items)
+        is SolidityBase.Array<*> -> value.items.map { Literal712(value = it).encodeData() }.reduce { acc, bytes -> acc + bytes }
+        is SolidityBase.Vector<*> -> value.items.map { Literal712(value = it).encodeData() }.reduce { acc, bytes -> acc + bytes }
+        else -> value.encode().hexToByteArray()
+    }
+
+    override val type = value.javaClass.simpleName.toLowerCase()
+}
+
+open class Struct712(override val type: String, val params: List<Pair<String, Type712>>) : Type712() {
```

Comment:
`Pair<String, Type712>` - I assume this is name to type, maybe adding a comment or a typealias would make this clearer

Reply:
Fixed in https://github.com/gnosis/svalinn-kotlin/pull/31/commits/d356f625de7cba047508c1d91dca99516e540c7f

Created a data class to have named parameters which I find cleaner what do you think?